library/alloc/src/sync.rs

```diff
@@ -3300,6 +3318,28 @@ impl<T: Default> Default for Arc<T> {
     }
 }
 
+/// Struct to hold the static `ArcInner` used for empty `Arc<str/CStr/[T]>` as
+/// returned by `Default::default`.
+///
+/// Layout notes:
+/// * `repr(align(16))` so we can use it for `[T]` with `align_of::<T>() <= 16`.
+/// * `repr(C)` so `inner` is at offset 0 (and thus guaranteed to actually be aligned to 16).
+/// * `[u8; 1]` (to be initialized with 0) so it can be used for `Arc<CStr>`.
+#[repr(C, align(16))]
+struct SliceArcInnerForStatic {
+    inner: ArcInner<[u8; 1]>,
+}
+const MAX_STATIC_INNER_SLICE_ALIGNMENT: usize = 16;
```

Comment:
This needs a `#[cfg(not(no_global_oom_handling))]`, since the `Default` impls using this only exist in that configuration.

Optionally we can put `#[cfg(not(no_global_oom_handling))]` on STATIC_INNER_SLICE too, and the debug-asserts in Drop, but it doesn't really hurt without.